include/taco/index_notation/provenance_graph.h

```diff
@@ -0,0 +1,362 @@
+class IndexVarRel : public util::IntrusivePtr<const IndexVarRelNode> {
+public:
+  IndexVarRel() : IntrusivePtr(nullptr) {}
+  IndexVarRel(IndexVarRelNode* node) : IntrusivePtr(node) {}
+  void print(std::ostream& stream) const;
+  bool equals(const IndexVarRel &rel) const;
+  IndexVarRelType getRelType() const;
+
+  template<typename T>
+  const T* getNode() const {
+    return static_cast<const T*>(ptr);
+  }
+
+  const IndexVarRelNode* getNode() const {
+    return ptr;
+  }
+};
+
+std::ostream& operator<<(std::ostream&, const IndexVarRel&);
+bool operator==(const IndexVarRel&, const IndexVarRel&);
+
+/// Index variable relations are used to track how new index variables are derived
+/// in the scheduling language
+struct IndexVarRelNode : public util::Manageable<IndexVarRelNode>,
+                         private util::Uncopyable {
+  IndexVarRelNode() : relType(UNDEFINED) {}
+  IndexVarRelNode(IndexVarRelType type) : relType(type) {}
+  virtual ~IndexVarRelNode() = default;
+  virtual void print(std::ostream& stream) const {
+    taco_iassert(relType == UNDEFINED);
+    stream << "underived";
+  }
+
+  /// returns list of index variables that are derived from
+  virtual std::vector<IndexVar> getParents() const {
+    taco_ierror;
+    return {};
+  }
+
+  /// returns list of index variables that are newly derived
+  virtual std::vector<IndexVar> getChildren() const {
+    taco_ierror;
+    return {};
+  }
+
+  /// returns list of index variables that are sized based on the size of the parent index variables (ie. outer for split)
+  virtual std::vector<IndexVar> getIrregulars() const {
+    taco_ierror;
+    return {};
+  }
+
+  /// Given coordinate bounds for parents, determine the new coordinate bounds relative to this possibly fused space
+  virtual std::vector<ir::Expr> computeRelativeBound(std::set<IndexVar> definedVars, std::map<IndexVar, std::vector<ir::Expr>> computedBounds, std::map<IndexVar, ir::Expr> variableExprs, Iterators iterators, ProvenanceGraph provGraph) const;
+
+  /// Determine the iteration bounds for a newly derived index variable given the iteration bounds of the parents
+  virtual std::vector<ir::Expr> deriveIterBounds(IndexVar indexVar, std::map<IndexVar, std::vector<ir::Expr>> parentIterBounds, std::map<IndexVar, std::vector<ir::Expr>> parentCoordBounds,
+                                                 std::map<taco::IndexVar, taco::ir::Expr> variableNames,
+                                                 Iterators iterators, ProvenanceGraph provGraph) const;
+
+  /// Recover a parent index variable expression as a function of the children index variables
+  virtual ir::Expr recoverVariable(IndexVar indexVar, std::map<IndexVar, ir::Expr> variableNames, Iterators iterators, std::map<IndexVar, std::vector<ir::Expr>> parentIterBounds, std::map<IndexVar, std::vector<ir::Expr>> parentCoordBounds, ProvenanceGraph provGraph) const;
```

Comment:
Some of these lines are very long. Would be nice to setup git clang format.

Reply:
I agree, I wish I was doing this throughout. I can run clang-format on this entire pull request, but am also nervous to make large changes now that Rawn and Tony have been working off this branch for several months. I will talk to them about it thanks.

Reply:
Yeah, I definitely didn't mean to suggest it should be done now. Maybe after the merge there could be a refactor commit on master along with some formatting guidelines? Clang format is pretty easy to setup, but I totally understand not wanting to throw in a whole bunch of changes while people are working.

Reply:
Ok I agree that would be a good way to go about it.